idlgen/src/type_names.rs

```diff
@@ -0,0 +1,188 @@
+        .type_params
+        .iter()
+        .find(|param| param.name == "E")
+        .ok_or_else(|| Error::TypeNotSupproted(format!("{type_info:?}")))?
+        .ty
+        .ok_or_else(|| Error::TypeNotSupproted(format!("{type_info:?}")))?;
+    let ok_type_name = resolve_type_name(type_registry, ok_type_id.id, resolved_type_names)?;
+    let err_type_name = resolve_type_name(type_registry, err_type_id.id, resolved_type_names)?;
+    Ok(format!("({ok_type_name}, {err_type_name})"))
+}
+
+fn option_type_name(
+    type_registry: &PortableRegistry,
+    type_info: &Type<PortableForm>,
+    resolved_type_names: &mut BTreeMap<u32, String>,
+) -> Result<String> {
+    let some_type_id = type_info
+        .type_params
+        .iter()
+        .find(|param| param.name == "T")
+        .ok_or_else(|| Error::TypeNotSupproted(format!("{type_info:?}")))?
+        .ty
+        .ok_or_else(|| Error::TypeNotSupproted(format!("{type_info:?}")))?;
+    let some_type_name = resolve_type_name(type_registry, some_type_id.id, resolved_type_names)?;
+    Ok(format!("opt {}", some_type_name))
+}
+
+fn type_name_by_path(type_info: &Type<PortableForm>) -> Result<String> {
+    let type_name = type_info
+        .path
+        .segments
+        .iter()
+        .map(|segment| segment.to_case(Case::Pascal))
+        .collect::<Vec<_>>()
+        .join("");
+    if type_name.is_empty() {
+        Err(Error::TypeNotSupproted(format!("{type_info:?}")))
+    } else {
+        Ok(type_name)
+    }
+}
+
+fn tuple_type_name(
+    type_registry: &PortableRegistry,
+    tuple_def: &TypeDefTuple<PortableForm>,
+    resolved_type_names: &mut BTreeMap<u32, String>,
+) -> Result<String> {
+    let fields = tuple_def
+        .fields
+        .iter()
+        .map(|field| resolve_type_name(type_registry, field.id, resolved_type_names))
+        .collect::<Result<Vec<_>>>()?
+        .join("; ");
+    if fields.is_empty() {
+        Ok("null".into()) // For the () type
+    } else {
+        Ok(format!("record {{ {} }}", fields))
+    }
+}
+
+fn array_type_name(
+    type_registry: &PortableRegistry,
+    item_type_id: u32,
+    resolved_type_names: &mut BTreeMap<u32, String>,
+) -> Result<String> {
+    let item_type_name = resolve_type_name(type_registry, item_type_id, resolved_type_names)?;
+    Ok(format!("vec {}", item_type_name))
+}
+
+fn primitive_type_name(type_def: &TypeDefPrimitive) -> Result<String> {
+    match type_def {
+        TypeDefPrimitive::Bool => Ok("bool".into()),
+        TypeDefPrimitive::Char => Ok("char".into()), // Candid doesn't have it. Do we want to support it? If such it will require a definition
+        TypeDefPrimitive::Str => Ok("text".into()),
+        TypeDefPrimitive::U8 => Ok("nat8".into()),
+        TypeDefPrimitive::U16 => Ok("nat16".into()),
+        TypeDefPrimitive::U32 => Ok("nat32".into()),
+        TypeDefPrimitive::U64 => Ok("nat64".into()),
+        TypeDefPrimitive::U128 => Ok("nat128".into()), // Candid doesn't have it
```

Comment:
Could be `[u8; 16]`, but candid doesn't have sized arrays either. Perhaps make it `vec nat8`? Not a blocker though

Reply:
I think this is more a question of how to represent it on a client side. Probably, that's the reason why Candid doesn't support it, and I'm leaning towards throwing it away for us too.

Reply:
Sized arrays are pretty much a question of the same matter - not all languages have a concept of fixed size arrays where the size is known at compile time.